File: apps/cli/src/lib/validate-skill-folder.ts
import { readdir, readFile, stat } from "node:fs/promises";
import { join, relative } from "node:path";

import { parseMentions } from "@better-skills/markdown/persisted-mentions";

import { collectNewResourceMentionPaths, normalizeResourcePath } from "./new-resource-mentions";

const RESOURCE_DIRS = ["references", "scripts", "assets"] as const;

export type SkillFolderValidationResult = {
  ok: boolean;
  errors: string[];
  warnings: string[];
  resourceCount: number;
  mentionCount: number;
};

function splitFrontmatter(skillMarkdown: string): { frontmatter: string | null; body: string } {
  if (!skillMarkdown.startsWith("---\n") && !skillMarkdown.startsWith("---\r\n")) {
    return { frontmatter: null, body: skillMarkdown };
  }

  const lines = skillMarkdown.split(/\r?\n/);
  let endIndex = -1;

  for (let i = 1; i < lines.length; i++) {
    if (lines[i]?.trim() === "---") {
      endIndex = i;
      break;
    }
  }

  if (endIndex === -1) {
    return { frontmatter: null, body: skillMarkdown };
  }

  return {
    frontmatter: lines.slice(1, endIndex).join("\n"),
    body: lines.slice(endIndex + 1).join("\n"),
  };
}

function hasFrontmatterField(frontmatter: string, field: "name" | "description"): boolean {
  const pattern = new RegExp(`^${field}:\\s*(.*)$`, "m");
  return pattern.test(frontmatter);
}

async function collectLocalResources(folder: string): Promise<Set<string>> {
  const resources = new Set<string>();

  for (const dirName of RESOURCE_DIRS) {
    const root = join(folder, dirName);
    const rootStat = await stat(root).catch(() => null);

    if (!rootStat?.isDirectory()) {
      continue;
    }

    const entries = await readdir(root, { recursive: true });

    for (const entry of entries) {
      const fullPath = join(root, entry);
      const fileStat = await stat(fullPath).catch(() => null);

      if (!fileStat?.isFile()) {
        continue;
      }

      const relativePath = normalizeResourcePath(relative(folder, fullPath));
      resources.add(relativePath);
    }
  }

  return resources;
}

type MentionCounts = {
  newPaths: string[];
  uuidResourceCount: number;
  total: number;
};

// We count two kinds of resource mentions to decide if every local file
// is accounted for:
//
//   1. [[resource:new:<path>]] — always internal, maps directly to a file
//   2. [[resource:<uuid>]]    — could be internal or external; we only
//      count it when the uuid belongs to this skill's own resources.
//      Clone writes a .resource-ids.json map so we can tell the difference.
//
// The total of (1) + (2) should equal the number of local resource files.
// If it doesn't, something is missing a mention.
async function collectAllResourceMentions(
  folder: string,
  skillMdBody: string,
  localResources: Set<string>,
): Promise<MentionCounts> {
  const newPathSet = new Set<string>();
  const uuidSet = new Set<string>();

  // load uuid→path map written by clone; absent for newly authored skills
  const resourceIds: Record<string, string> = await readFile(
    join(folder, ".resource-ids.json"),
    "utf8",
  )
    .then((raw) => JSON.parse(raw))
    .catch(() => ({}));
  const internalIds = new Set(Object.keys(resourceIds).map((id) => id.toLowerCase()));

  const scanMarkdown = (text: string) => {
    for (const p of collectNewResourceMentionPaths(text)) newPathSet.add(p);
    for (const m of parseMentions(text)) {
      // skip external resource mentions and skill mentions
      if (m.type !== "resource") continue;
      if (internalIds.size > 0 && !internalIds.has(m.targetId)) continue;
      uuidSet.add(m.targetId);
    }
  };

  scanMarkdown(skillMdBody);

  for (const resourcePath of localResources) {
    if (!resourcePath.endsWith(".md")) continue;
    const content = await readFile(join(folder, resourcePath), "utf8").catch(() => null);
    if (!content) continue;
    scanMarkdown(content);
  }

  const newPaths = [...newPathSet];
  return {
    newPaths,
    uuidResourceCount: uuidSet.size,
    total: newPathSet.size + uuidSet.size,
  };
}

export async function validateSkillFolder(folder: string): Promise<SkillFolderValidationResult> {
  const errors: string[] = [];
  const warnings: string[] = [];

  const folderStat = await stat(folder).catch(() => null);
  if (!folderStat?.isDirectory()) {
    return {
      ok: false,
      errors: [`not a directory: ${folder}`],
      warnings,
      resourceCount: 0,
      mentionCount: 0,
    };
  }

  const skillMdPath = join(folder, "SKILL.md");
  const content = await readFile(skillMdPath, "utf8").catch(() => null);

  if (!content) {
    return {
      ok: false,
      errors: ["SKILL.md not found"],
      warnings,
      resourceCount: 0,
      mentionCount: 0,
    };
  }

  const { frontmatter, body } = splitFrontmatter(content);

  if (frontmatter === null) {
    errors.push("SKILL.md must start with YAML frontmatter delimited by ---");
  } else {
    if (!hasFrontmatterField(frontmatter, "name")) {
      errors.push("frontmatter missing required field: name");
    }

    if (!hasFrontmatterField(frontmatter, "description")) {
      errors.push("frontmatter missing required field: description");
    }
  }

  const localResources = await collectLocalResources(folder);
  const mentions = await collectAllResourceMentions(folder, body, localResources);

  const missing = mentions.newPaths.filter((path) => !localResources.has(path));
  if (missing.length > 0) {
    errors.push("missing local resources for :new: mention tokens:");
    errors.push(...missing.map((path) => `  - ${path}`));
  }

  if (localResources.size > 0 && mentions.total < localResources.size) {
    const uncovered = localResources.size - mentions.total;

    if (mentions.uuidResourceCount === 0) {
      // all mentions are :new: — we know exactly which files are uncovered
      const newPathSet = new Set(mentions.newPaths);
      const unreferenced = [...localResources].filter((path) => !newPathSet.has(path));
      warnings.push(
        [
          `${uncovered} resource file(s) not referenced by any mention:`,
          ...unreferenced.map((path) => `  - ${path}`),
        ].join("\n"),
      );
    } else {
      warnings.push(`${uncovered} resource file(s) not referenced by any mention`);
    }
  }

  return {
    ok: errors.length === 0,
    errors,
    warnings,
    resourceCount: localResources.size,
    mentionCount: mentions.total,
  };
}

export function formatValidationFailure(result: SkillFolderValidationResult): string {
  const lines = ["validation failed", ...result.errors.map((error) => `- ${error}`)];

  if (result.warnings.length > 0) {
    lines.push(...result.warnings.map((warning) => `- warning: ${warning}`));
  }

  return lines.join("\n");
}

export async function assertValidSkillFolder(folder: string): Promise<SkillFolderValidationResult> {
  const result = await validateSkillFolder(folder);

  if (!result.ok || result.warnings.length > 0) {
    throw new Error(formatValidationFailure(result));
  }

  return result;
}

File: resources/default-skills/better-skills/scripts/validate_skill_folder.py
#!/usr/bin/env python3

"""Validate a better-skills local skill folder before create/update.

Checks:
- SKILL.md exists
- frontmatter contains name + description
- each [[skill:new:...]] / [[resource:new:...]] path exists locally
"""

from __future__ import annotations

import argparse
import re
import sys
from pathlib import Path


NEW_MENTION_RE = re.compile(
    r"(?<!\\)\[\[(skill|resource):new:([^\]\n]+)\]\]", re.IGNORECASE
)


def normalize_path(path: str) -> str:
    normalized = path.strip().replace("\\", "/")

    if normalized.startswith("./"):
        normalized = normalized[2:]

    if "#" in normalized:
        normalized = normalized.split("#", 1)[0]

    if "?" in normalized:
        normalized = normalized.split("?", 1)[0]

    while normalized.startswith("/"):
        normalized = normalized[1:]

    return normalized


def split_frontmatter(skill_md: str) -> tuple[str | None, str]:
    if not skill_md.startswith("---\n"):
        return None, skill_md

    lines = skill_md.splitlines()
    end_idx = None

    for i in range(1, len(lines)):
        if lines[i].strip() == "---":
            end_idx = i
            break

    if end_idx is None:
        return None, skill_md

    frontmatter = "\n".join(lines[1:end_idx])
    body = "\n".join(lines[end_idx + 1 :])
    return frontmatter, body


def has_frontmatter_field(frontmatter: str, field: str) -> bool:
    pattern = re.compile(rf"(?m)^{re.escape(field)}:\s*(.*)$")
    return bool(pattern.search(frontmatter))


def collect_local_resources(folder: Path) -> set[str]:
    paths: set[str] = set()

    for root in ("references", "scripts", "assets"):
        root_dir = folder / root
        if not root_dir.exists() or not root_dir.is_dir():
            continue

        for file_path in root_dir.rglob("*"):
            if file_path.is_file():
                rel = file_path.relative_to(folder).as_posix()
                paths.add(normalize_path(rel))

    return paths


def collect_new_mention_paths(markdown: str) -> list[str]:
    seen: set[str] = set()
    result: list[str] = []

    in_fence = False
    fence_marker: str | None = None

    for line in markdown.splitlines():
        stripped = line.lstrip()
        fence_match = re.match(r"(```+|~~~+)", stripped)

        if fence_match:
            marker = fence_match.group(1)[0]
            if not in_fence:
                in_fence = True
                fence_marker = marker
            elif fence_marker == marker:
                in_fence = False
                fence_marker = None
            continue

        if in_fence:
            continue

        parts = re.split(r"(`[^`]*`)", line)
        for part in parts:
            if len(part) >= 2 and part.startswith("`") and part.endswith("`"):
                continue

            for match in NEW_MENTION_RE.finditer(part):
                path = normalize_path(match.group(2))
                if path not in seen:
                    seen.add(path)
                    result.append(path)

    return result


def validate(folder: Path) -> int:
    errors: list[str] = []
    warnings: list[str] = []

    skill_file = folder / "SKILL.md"
    if not skill_file.exists() or not skill_file.is_file():
        print("error: SKILL.md not found")
        return 1

    content = skill_file.read_text(encoding="utf-8")
    frontmatter, body = split_frontmatter(content)

    if frontmatter is None:
        errors.append("SKILL.md must start with YAML frontmatter delimited by ---")
    else:
        if not has_frontmatter_field(frontmatter, "name"):
            errors.append("frontmatter missing required field: name")
        if not has_frontmatter_field(frontmatter, "description"):
            errors.append("frontmatter missing required field: description")

    local_resources = collect_local_resources(folder)
    mention_paths = collect_new_mention_paths(body)

    missing = [path for path in mention_paths if path not in local_resources]
    if missing:
        errors.append("missing local resources for :new: mention tokens:")
        errors.extend([f"  - {path}" for path in missing])

    if not local_resources:
        warnings.append(
            "no local resources found under references/, scripts/, or assets/"
        )

    if errors:
        print("validation failed")
        for line in errors:
            print(f"- {line}")
        if warnings:
            for line in warnings:
                print(f"- warning: {line}")
        return 1

    print("validation passed")
    print(f"- resources: {len(local_resources)}")
    print(f"- :new: mention paths: {len(mention_paths)}")
    if warnings:
        for line in warnings:
            print(f"- warning: {line}")
    return 0


def main() -> int:
    parser = argparse.ArgumentParser(
        description="Validate Better Skills local skill folder"
    )
    parser.add_argument("folder", help="Path to local skill folder")
    args = parser.parse_args()

    folder = Path(args.folder).expanduser().resolve()
    if not folder.exists() or not folder.is_dir():
        print(f"error: not a directory: {folder}")
        return 1

    return validate(folder)


if __name__ == "__main__":
    sys.exit(main())
